superset-frontend/spec/javascripts/dashboard/util/getFilterScopeFromNodesTree_spec.js

```diff
@@ -157,6 +159,107 @@ describe('getFilterScopeFromNodesTree', () => {
       },
     ];
 
+    // this is an other commonly used layout for dashboard:
```

Comment:
Would be nice if there's a short note that visualizes the dashboard layout:

```js
// - filter0
// - tab1
//   - filter1
//   - chart1
// - tab2
//   - filter2 
//   - chart2
//   - tab2-1
//     - chart3
//     - chart4
//   - tab2-2
//     - chart5
//     - chart6
//
// The test case expects it's possible to apply filter2 to filter0
```

On a separate note, `chartsImmune` should probably be refactored to `chartsInScope` so the data structure can better align with the UI . Negative condition sometimes is difficult for a human brain to comprehend. ESLint has a rule for it: https://eslint.org/docs/rules/no-negated-condition . (Not that we need to do it soon, of course, but something to keep in mind when it's time to revamp the filters)

Reply:
Dashboard filter's `scope` data is defined as 
- `scopes`: contain one or more tabs(or root_id when dashboard has no tab). 
- `immune` are charts under tab but immune from filter.
If we only hold a list of `chartsInScope` will lost tab's scope information.

Reply:
You have two options:

1. Never store tab ids in scopes, but rather all charts in scope for that tab. All new charts has all filters turned on by default.
2. Change "immune" to `tabScopes`. If tab id is new `scopes`, then new charts added to tab will apply the filter by default, otherwise not.

It's probably better to just have a dict of flat arrays `{ [FilterId]: ChartId[] }`, in which empty means applicable to all charts, and non-empty means only apply to selected charts. Tab tree status can be computed on the fly.